src/app/qgsattributetabledialog.cpp

```diff
@@ -217,16 +217,18 @@ QgsAttributeTableDialog::QgsAttributeTableDialog( QgsVectorLayer *layer, QgsAttr
   {
     request.setFilterExpression( filterExpression );
   }
-  if ( !needsGeom )
-    request.setFlags( Qgis::FeatureRequestFlag::NoGeometry );
 
+  if ( !needsGeom && ( !mLayer || !QgsExpression( mLayer->attributeTableConfig().sortExpression() ).needsGeometry() ) )
+  {
+    request.setFlags( Qgis::FeatureRequestFlag::NoGeometry );
+  }
```

Comment:
I'd find this slightly more readable:

```suggestion
  // If sort expression requires geometry, we'll need to fetch it
  needsGeom |= mLayer && QgsExpression( mLayer->attributeTableConfig().sortExpression() ).needsGeometry();  
  if ( !needsGeom )
    request.setFlags( Qgis::FeatureRequestFlag::NoGeometry );
```

Reply:
Or `needsGeom = needsGeom || ( mLayer && QgsExpression( mLayer->attributeTableConfig().sortExpression() ).needsGeometry() );` to short-circuit when needsGeom is already True?

Reply:
That's exactly the same as `|=`

Reply:
Sorry, I supposed `x |= y;` doesn't short-circuit like `x = x || y();` so `y` was always evaluated even if `x` is already true.